src/fibsem_tools/io/h5.py

```diff
@@ -0,0 +1,57 @@
+import h5py
+from pathlib import Path
+from typing import Union, Any, Dict
+import warnings
+
+Pathlike = Union[str, Path]
+
+
+def partition_h5_kwargs(**kwargs) -> Dict[str, Any]:
+    """
+    partition kwargs into file-creation kwargs and dataset-creation kwargs
+    """
+    file_kwargs = kwargs.copy()
+    dataset_kwargs = {}
+    for key in ("shape", "dtype", "data", "chunks", "compressor"):
+        if key in file_kwargs:
+            dataset_kwargs[key] = file_kwargs.pop(key)
+
+    if (data := kwargs.get("data")) is not None:
+        if "dtype" in kwargs and (dtype := data.dtype) != kwargs["dtype"]:
+            warnings.warn(
+                f"{dtype=} and {data=} were provided as a keyword arguments but they conflict. dtype will be inferred from `data`"
+            )
+        dataset_kwargs["dtype"] = data.dtype
+
+        if "shape" in kwargs and (shape := data.shape) != kwargs["shape"]:
+            warnings.warn(
+                f"{shape=} and {data=} were provided as a keyword arguments but they conflict. shape will be inferred from `data`"
+            )
+
+        dataset_kwargs["shape"] = data.shape
+
+    return file_kwargs, dataset_kwargs
+
+
+def access_h5(
+    store: Pathlike, path: Pathlike, mode: str, **kwargs
```

Comment:
How about allowing `store` to be an open `h5py.File` handle. If it is, then the user manages the file context. If it is not, then this function handles closing the file?

Reply:
I played around with this (and I see the appeal) but it complicates the function signature of `access_h5` -- if you allow the first argument to be an already-opened `h5py.File` object, then the `mode` argument to `access_h5` becomes meaningless and would need to be ignored. I'm not happy when my function arguments have dependencies like this, and `access_h5` is a very light wrapper, so I decided not to overload the first argument.

Reply:
Shouldn't `mode` be a keyword argument like in the Zarr version? The `mode` keyword has a default as `'r'` upstream. Why is it required here and why is it different than the other keyword arguments?

https://github.com/h5py/h5py/blob/50b6dc1c9208614c3dbba5ecc5f4d7607525fcd7/h5py/_hl/files.py#L363-L367

If someone supplies both a `h5py.File` and a `mode` keyword or any of the file keywords, that should be an error. Alternatively you could assert that the `mode` property of the `h5py.File` instance is the same as the mode keyword argument.

Reply:
yeah, I think that could work, and +1 to the idea of normalizing the API to match `access_zarr`